### PR TITLE
documentation for new builder methods

### PR DESCRIPTION
**Pull request by HEskandari**

### CODE REVIEW

File: Snippets/ASBFunctions/ASBFunctions_1.3/FunctionsHostBuilderUsage.cs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+// needs to be top level first defined for C# to compile
+#region endpoint-trigger-function-wireup
+
+[assembly: NServiceBusTriggerFunction("MyFunctionsEndpoint",
+    TriggerFunctionName = "NServiceBusFunctionEndpointTrigger-ASBTriggerQueue")]
+
+#endregion
+
+namespace ASBFunctions_1_2_Autogenerated
```
Comment (mikeminutillo) [inline Snippets/ASBFunctions/ASBFunctions_1.3/FunctionsHostBuilderUsage.cs:11]: Should these namespaces be 1_3?

File: nservicebus/hosting/azure-functions/service-bus_configuration_ASBFunctions_[1.3,).partial.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+NServiceBus can be registered and configured on the host builder using the `UseNServiceBus` extension method in the startup class:
+
+snippet: asb-function-default
+
+The default builder method will look up values such as function name, connection string and license information directly from from the `IConfiguration`. Fine-grain configuration is available through the overload accepting a `ServiceBusTriggeredEndpointConfiguration` class.
```
Comment (mikeminutillo) [inline nservicebus/hosting/azure-functions/service-bus_configuration_ASBFunctions_[1.3,).partial.md:5]: The overload doesn't accept a `ServiceBusTriggeredEndpointConfiguration`. It accepts a factory method to create one.

We may be able to simplify things by just saying fine-grained config is also possible and then putting the asb-function-hostbuilder snippet in.

File: Snippets/ASBFunctions/ASBFunctions_1.3/FakeFunctionsSdk.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+public class HttpRequest
+{
+}
+
+enum AuthorizationLevel
+{
+    Function
+}
```
Comment (mikeminutillo) [inline Snippets/ASBFunctions/ASBFunctions_1.3/FakeFunctionsSdk.cs:31]: Why do this and not just reference the real sdk?

File: nservicebus/hosting/azure-functions/service-bus_queue-trigger-wiring_ASBFunctions_[1.2,).partial.md [hunk 4]
```diff
@@ -1,19 +0,0 @@
...
-
-If the trigger function needs to be customized, the trigger function generation can be disabled by removing the `NServiceBusEndpointName` attribute. A customized trigger function can then be manually added to the project.
-
-#### Overriding the trigger function name
-
-The trigger function name is auto-generated by default. To customize the function name, the `NServiceBusEndpointName` attribute can be provided with an additional parameter to set the function name:
-
-snippet: asb-function-override-trigger-function-name
-
-Note: An invalid trigger function name will generate an `NSBFUNC002` error with the message `Trigger function name is invalid and cannot be used to generate trigger function`.
```
Comment (mikeminutillo) [inline nservicebus/hosting/azure-functions/service-bus_queue-trigger-wiring_ASBFunctions_[1.2,).partial.md:19]: Why is this removed? Are these snippets no longer used? When I render this locally, the 1.2 page has a heading with no content for this:

image

File: nservicebus/hosting/azure-functions/service-bus_configuration_ASBFunctions_[,1.2].partial.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+NServiceBus can be registered and configured on the host builder using the `UseNServiceBus` extension method in the startup class:
```
Comment (mikeminutillo) [inline nservicebus/hosting/azure-functions/service-bus_configuration_ASBFunctions_[,1.2].partial.md:1]: As far as I can tell, this is the same in both partials. That's not really an issue but it could be pulled up into the main file,

File: Snippets/ASBFunctions/ASBFunctions_1.3/FakeFunctionsSdk.cs (hunk 3 above)
Comment (HEskandari) [reply Snippets/ASBFunctions/ASBFunctions_1.3/FakeFunctionsSdk.cs:31]: don't know exactly...possibly because we didn't want to install other SKDs (Azure etc) on the build engine? The older versions are also using stubs like this.

File: nservicebus/hosting/azure-functions/service-bus_queue-trigger-wiring_ASBFunctions_[1.2,).partial.md (hunk 4 above)
Comment (HEskandari) [reply nservicebus/hosting/azure-functions/service-bus_queue-trigger-wiring_ASBFunctions_[1.2,).partial.md:19]: fixed, the md file was deleted causing issues.

File: Snippets/ASBFunctions/ASBFunctions_1.3/FunctionsHostBuilderUsage.cs (hunk 1 above)
Comment (HEskandari) [reply Snippets/ASBFunctions/ASBFunctions_1.3/FunctionsHostBuilderUsage.cs:11]: updated

File: nservicebus/hosting/azure-functions/service-bus_configuration_ASBFunctions_[1.3,).partial.md (hunk 2 above)
Comment (HEskandari) [reply nservicebus/hosting/azure-functions/service-bus_configuration_ASBFunctions_[1.3,).partial.md]: Rephrased. See if it is better now?

File: nservicebus/hosting/azure-functions/service-bus_configuration_ASBFunctions_[,1.2].partial.md (hunk 5 above)
Comment (HEskandari) [reply nservicebus/hosting/azure-functions/service-bus_configuration_ASBFunctions_[,1.2].partial.md]: moved to the parent file.

File: Snippets/ASBFunctions/ASBFunctions_1.3/FakeFunctionsSdk.cs (hunk 3 above)
Comment (timbussmann) [reply Snippets/ASBFunctions/ASBFunctions_1.3/FakeFunctionsSdk.cs:31]: if it's possible, using the real SDK seems to make more sense unless that breaks something?

File: Snippets/ASBFunctions/ASBFunctions_1.3/FunctionsHostBuilderUsage.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
+using NServiceBus;
+
+// needs to be top level first defined for C# to compile
+#region endpoint-trigger-function-wireup
+
+[assembly: NServiceBusTriggerFunction("MyFunctionsEndpoint",
+    TriggerFunctionName = "NServiceBusFunctionEndpointTrigger-ASBTriggerQueue")]
```
Comment (timbussmann) [inline Snippets/ASBFunctions/ASBFunctions_1.3/FunctionsHostBuilderUsage.cs:7]: any reason that we specifically define the function name?

File: Snippets/ASBFunctions/ASBFunctions_1.3/FunctionsHostBuilderUsage.cs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    using Microsoft.Azure.ServiceBus.Core;
+    using Microsoft.Azure.WebJobs;
+    using Microsoft.Extensions.Logging;
+    using NServiceBus;
+
+    class FunctionEndpointTrigger
+    {
+        readonly IFunctionEndpoint endpoint;
+
+        public FunctionEndpointTrigger(IFunctionEndpoint endpoint)
```
Comment (timbussmann) [inline Snippets/ASBFunctions/ASBFunctions_1.3/FunctionsHostBuilderUsage.cs:26]: that's not what the new generator generates, as it uses `FunctionEndpoint` directly. Not sure that's something we should show as we want users to use the `IFunctionEndpoint` for manual usage. Do we even need to show the generated code at all?

File: Snippets/ASBFunctions/ASBFunctions_1.3/FunctionsHostBuilderUsage.cs [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+                    return configuration;
+                });
+            }
+        }
+        #endregion
+
+        #region asb-enable-diagnostics
+        class configureErrorQueueuOnStartup : FunctionsStartup
+        {
+            public override void Configure(IFunctionsHostBuilder builder)
```
Comment (timbussmann) [inline Snippets/ASBFunctions/ASBFunctions_1.3/FunctionsHostBuilderUsage.cs:103]: could we move the region before the `Configure` methods to leave out the class definition on the snippets?

File: Snippets/ASBFunctions/ASBFunctions_1.3/FunctionsHostBuilderUsage.cs (hunk 7 above)
Comment (HEskandari) [reply Snippets/ASBFunctions/ASBFunctions_1.3/FunctionsHostBuilderUsage.cs:26]: removed the auto-genned section from the snippet.

File: Snippets/ASBFunctions/ASBFunctions_1.3/FunctionsHostBuilderUsage.cs (hunk 6 above)
Comment (HEskandari) [reply Snippets/ASBFunctions/ASBFunctions_1.3/FunctionsHostBuilderUsage.cs:7]: no reason, just copied code from the earlier version. I'll remove.

File: Snippets/ASBFunctions/ASBFunctions_1.3/FakeFunctionsSdk.cs (hunk 3 above)
Comment (HEskandari) [reply Snippets/ASBFunctions/ASBFunctions_1.3/FakeFunctionsSdk.cs]: used the SDK and removed all the stubs.

**Review by timbussmann — APPROVED** [review]
